mava/systems/q_learning/rec_iql.py

```diff
@@ -0,0 +1,644 @@
+        buffer_state,
+        params,
+        first_keys,
+    )
+
+    return (env, eval_env), q_net, opt, rb, learner_state, logger, key
+
+
+def make_update_fns(
+    cfg: DictConfig,
+    env: Environment,
+    q_net: RecQNetwork,
+    opt: optax.GradientTransformation,
+    rb: TrajectoryBuffer,
+) -> Callable[[LearnerState], Tuple[LearnerState, Tuple[Metrics, Metrics]]]:
+    """Create the update function for the Q-learner.
+
+    Args:
+        cfg: System configuration.
+        env: Learning environment.
+        q_net: Recurrent q network.
+        opt: Optimiser for the recurrent Q network.
+        rb: The replay buffer.
+
+    Returns:
+        The update function.
+    """
+
+    # INTERACT LEVEL 2
+    def select_eps_greedy_action(
+        action_selection_state: ActionSelectionState, obs: Observation, done: Array
+    ) -> Tuple[ActionSelectionState, Array]:
+        """Select action to take in epsilon-greedy way. Batch and agent dims are included.
+
+            Args:
+            action_selection_state: Tuple of online parameters, previous hidden state,
+                environment timestep (used to calculate epsilon) and a random key.
+            obs: The observation from the previous timestep.
+            done: The flag timestep.last() from the previous timestep.
+
+        Returns:
+            A tuple of the updated action selection state and the chosen action.
+        """
+        params, hidden_state, t, key = action_selection_state
+
+        eps = jax.numpy.maximum(
+            cfg.system.eps_min, 1 - (t / cfg.system.eps_decay) * (1 - cfg.system.eps_min)
+        )
+
+        obs = jax.tree_util.tree_map(lambda x: x[jnp.newaxis, ...], obs)
+        done = jax.tree_util.tree_map(lambda x: x[jnp.newaxis, ...], done)
+
+        next_hidden_state, eps_greedy_dist = q_net.apply(params, hidden_state, (obs, done), eps)
+
+        new_key, explore_key = jax.random.split(key, 2)
+
+        action = eps_greedy_dist.sample(seed=explore_key)
+        action = jnp.squeeze(action)
+
+        next_action_selection_state = ActionSelectionState(
+            params, next_hidden_state, t + cfg.arch.num_envs, new_key
+        )
+
+        return next_action_selection_state, action
+
+    # INTERACT LEVEL 1
+    def interaction_step(
+        interaction_state: InteractionState, _: Any
+    ) -> Tuple[InteractionState, Dict]:
+        """Selects action, steps global env, stores timesteps in rb and repacks the parameters."""
+        # light unpacking
+        action_selection_state, env_state, buffer_state, obs, done = interaction_state
+
+        # select the actions to take
+        next_action_selection_state, action = select_eps_greedy_action(
+            action_selection_state, obs, done
+        )
+
+        # step env with selected actions
+        next_env_state, next_timestep = jax.vmap(env.step)(env_state, action)
+
+        # Get reward
+        reward = jnp.mean(
+            next_timestep.reward, axis=-1, keepdims=True
```

Comment:
The algorithm does not currently train using individual rewards. For VDN, it also makes sense to combine rewards. If we need to be able to use individual rewards, let's dicuss!

Reply:
I think that maybe we should actually make it work with individual rewards so that you could in principle run it on an environment with independent rewards.

Reply:
Will do! I am going to work on making it an option in the config.